bitvm/src/bn254/g2_sg_check.rs

```diff
@@ -0,0 +1,704 @@
+use std::{ops::Neg, str::FromStr};
+
+use ark_ec::{AffineRepr, CurveGroup};
+use ark_ff::{AdditiveGroup, Field};
+use bitcoin_script::script;
+use num_bigint::BigUint;
+use crate::treepp::Script;
+
+use super::{fp254impl::Fp254Impl, fq::Fq, fq2::Fq2, g2::{hinted_affine_add_line, hinted_affine_double_line, hinted_check_chord_line, hinted_check_tangent_line, G2Affine}, utils::Hint};
+
+fn split_scalar(window: usize, scalar: u64) -> Vec<Vec<u8>> {
+    fn u64_to_bits(x: u64) -> Vec<u8> {
+        let mut bits = Vec::with_capacity(64);
+        for i in 0..64 {
+            // Shift so that we're checking the (63 - i)-th bit from the right.
+            // That puts the most significant bit at i = 0.
+            let bit = ((x >> (63 - i)) & 1) as u8;
+            if bit == 0 {
+                bits.push(0); // dbl
+            } else if bit == 1 {
+                bits.push(0); // dbl
+                bits.push(1); // add
+            }
+        }
+        bits
+    }
+    let mut scalar_bits: Vec<Vec<u8>> = vec![];
+    u64_to_bits(scalar).chunks(window as usize).for_each(|c| {
+        scalar_bits.push(c.to_vec());
+    });
+    scalar_bits
+}
+
+fn hinted_check_double_and_add(t: ark_bn254::G2Affine, q: ark_bn254::G2Affine, bits: Vec<u8>) -> (Script, Vec<Hint>) {
+    let mut hints: Vec<Hint> = vec![];
+    let mut acc = t.clone();
+    let mut script = script!();
+    for bit in bits {
+        if bit == 0 {
+            let (scr, hint) = hinted_check_double(acc);
+            hints.extend_from_slice(&hint);
+            script = script!(
+                {script}
+                {Fq2::toaltstack()} {Fq2::toaltstack()} // move q to altstack
+                {scr}
+                {Fq2::fromaltstack()} {Fq2::fromaltstack()} // bring q to altstack
+            );
+            acc = (acc + acc).into_affine(); // double
+            
+        } else if bit == 1 {
+            let (scr, hint) = hinted_check_add(acc, q);
+            hints.extend_from_slice(&hint);
+            script = script!(
+                {script}
+                {Fq2::copy(2)} {Fq2::copy(2)}
+                {Fq2::toaltstack()} {Fq2::toaltstack()} // move q to altstack
+                {scr}
+                {Fq2::fromaltstack()} {Fq2::fromaltstack()} // bring q to altstack
```

Comment:
Same here.

Reply:
added understandable comment showing stack snapshot in latest commit.

Reply:
Thank you, even better!